feat: add Prune and Delete as application level sync option

MrFreezeex opened the pull request:
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

My goal was mainly to add Prune/Delete=confirm as a sync option as the deletion confirmation acts on the whole Application already and it's way more convenient in our env to add it on the application level than as annotation on the resources actually deployed.

For consistency I also added Prune/Delete=false support there (separate commit).

And the UI option in when editing the sync policy for the Prune and Delete sync options

Closes https://github.com/argoproj/argo-cd/issues/23380

pjiang-dev reviewed (commented):
Could you create an issue describing what the issue/enhancement is and why this change is needed and link it to this PR and your gitops-engine PR?

pjiang-dev commented:
The deletion part LGTM. Maybe helpful to add a screenshot or video of the sync option working. Thanks

MrFreezeex commented on docs/user-guide/sync-options.md:
````diff
@@ -14,6 +14,17 @@ metadata:
     argocd.argoproj.io/sync-options: Prune=false
 ```
 
+It also can be enabled at the application level like in the example below:
````
Done! Let me know if the new doc is more clear!

MrFreezeex commented:
@pjiang-dev here is the video (I had to fix a dumb typo https://github.com/argoproj/gitops-engine/compare/13f9343d366f59f29905dda9c8ef4e6257a67c67..03789b820ebec73de41af43a35493983c1832e01 in the gitops-engine so thanks for making me double check :see_no_evil:)


https://github.com/user-attachments/assets/4cad277a-b821-46cb-94e8-c28cc1235a82

MrFreezeex commented:
FYI here is the yaml that I am using in the video above:
```
---
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: test-prune-confirm
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  project: default
  source:
    repoURL: https://github.com/MrFreezeex/test-argocd-confirm.git
    targetRevision: "testsyncopt"
    path: test-prune
    # path: test-empty
  destination:
    name: in-cluster
    namespace: test-prune-confirm
  syncPolicy:
    automated: {}
    syncOptions:
      - Prune=confirm
---
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: test-prune-false
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  project: default
  source:
    repoURL: https://github.com/MrFreezeex/test-argocd-confirm.git
    targetRevision: "testsyncopt"
    path: test-prune
    # path: test-empty
  destination:
    name: in-cluster
    namespace: test-prune-false
  syncPolicy:
    automated: {}
    syncOptions:
      - Prune=false
---
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: test-delete-confirm
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  project: default
  source:
    repoURL: https://github.com/MrFreezeex/test-argocd-confirm.git
    targetRevision: "testsyncopt"
    path: test-delete
  destination:
    name: in-cluster
    namespace: test-delete-confirm
  syncPolicy:
    automated: {}
    syncOptions:
      - Delete=confirm
---
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: test-delete-false
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  project: default
  source:
    repoURL: https://github.com/MrFreezeex/test-argocd-confirm.git
    targetRevision: "testsyncopt"
    path: test-delete
  destination:
    name: in-cluster
    namespace: test-delete-false
  syncPolicy:
    automated: {}
    syncOptions:
      - Delete=false
```

agaudreault commented on docs/user-guide/sync-options.md:
````diff
@@ -34,6 +49,20 @@ metadata:
 To confirm the pruning you can use Argo CD UI, CLI or manually apply the `argocd.argoproj.io/deletion-approved: <ISO formatted timestamp>`
 annotation to the application.
 
+It also can be enabled at the application level like in the example below:
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+spec:
+  syncPolicy:
+    syncOptions:
+    - Prune=confirm
+```
+
+Note that setting a Prune sync option on the application level will not override
+the same sync option set on a specific resource, both will still be applied.
+
````
Instead of repeating the doc, specify that the same value can be use either through the annotation or the sync option.

MrFreezeex replied:
I don't mind doing that but IIUC adding both like that is consistent with the current doc, for instance ServerSideApply: https://argo-cd.readthedocs.io/en/stable/user-guide/sync-options/#server-side-apply

agaudreault commented on ui/src/app/applications/components/application-sync-options/application-sync-options.tsx:
```diff
@@ -113,6 +113,8 @@ export const ApplicationSyncOptions = (props: ApplicationSyncOptionProps) => (
         ))}
         <div className='small-12' style={optionStyle}>
             {booleanOption('Replace', 'Replace', false, props, false, REPLACE_WARNING)}
+            {selectOption('Prune', 'Prune', 'true', ['true', 'false', 'confirm'], props)}
```
Should reorder items so Prune, Prune Last and PrunePropagationPolicy are next to each other.

The UI should also group the boolean options together, then the selectOptions.

agaudreault commented on test/e2e/app_management_test.go:
```diff
@@ -1747,6 +1747,32 @@ func TestSyncOptionPruneFalse(t *testing.T) {
 }
 
+// make sure that if we deleted a resource from the app, it is not pruned if the app has the sync option Prune=false
+func TestSyncOptionSyncOptionPruneFalse(t *testing.T) {
```
add e2e test when option is specified at the application level and resource level

MrFreezeex commented:
Hi @agaudreault, kind reminder about this PR, would love to have your update input on the matter whenever you have time do so :pray:

sharovmerk commented:
+1 I really need it

agaudreault commented on docs/user-guide/sync-options.md:
````diff
@@ -14,6 +14,21 @@ metadata:
+    - Prune=false
+```
+
+Note that setting a Prune sync option on the application level will not override
+the same sync option set on a specific resource, both will still be applied.
````
I think it is important that the resource can override the setting in the Applications. I think it is actually safer, because you can have mutationWebhooks adding these annotations to specific resources. If an annotation is specified on a resource, the Application level setting should not take precedence.

agaudreault reviewed (changes requested):
Waiting for changes over the resource annotation precedence. Other than that, feature looks good 👍

MrFreezeex committed:
chore: add utils to get sync option value

This will facilitate adding logic on sync options that can be defined
both at the app and resource level and the resource level should
override the app level option.

Also fix GetAnnotationCSVs to preserve the order returned (as map in
golang doesn't preserve the insertion order) to make dedup works in
GetOptionValue.

Signed-off-by: Arthur Outhenin-Chalandre <arthur.outhenin-chalandre@ledger.fr>

MrFreezeex committed:
feat: add Delete/Prune=confirm sync option on application

Add support for Delete and Prune sync options on the application level.
When at the application level those options are only changing the
"default behavior" for resources deployed by this application.

The behavior is still to process this at the resource level and the resource
level always override the app level Delete option.

Signed-off-by: Arthur Outhenin-Chalandre <arthur.outhenin-chalandre@ledger.fr>

MrFreezeex committed:
chore: add e2e test for Delete/Prune option on application

Signed-off-by: Arthur Outhenin-Chalandre <arthur.outhenin-chalandre@ledger.fr>

MrFreezeex committed:
feat: add Prune and delete sync option edit on UI

Also reorder the PruneLast option so that all prune options are grouped

Signed-off-by: Arthur Outhenin-Chalandre <arthur.outhenin-chalandre@ledger.fr>

agaudreault commented on gitops-engine/pkg/sync/sync_context.go:
```diff
@@ -429,7 +437,14 @@ func (sc *syncContext) setRunningPhase(tasks syncTasks, isPendingDeletion bool)
+			if pruneOptionValue == nil {
+				pruneOptionValue = sc.defaultPruneOption
+			}
+			return task.isPrune() && pruneOptionValue != nil && *pruneOptionValue == common.SyncValueConfirm
```
Extract to an objRequiresPruneConfirmation method


```suggestion
			return task.isPrune() && objRequiresPruneConfirmation(task.liveObj, sc.defaultPruneOption)
```

agaudreault commented on gitops-engine/pkg/sync/sync_context.go:
```diff
@@ -1389,7 +1404,13 @@ func (sc *syncContext) applyObject(t *syncTask, dryRun, validate bool) (common.R
+	if pruneOptionValue == nil {
+		pruneOptionValue = sc.defaultPruneOption
+	}
+	if pruneOptionValue != nil && *pruneOptionValue == common.SyncValueFalse {
```
Extract to a isPruningDisabled() method.
```suggestion
	if isPruningDisabled(liveObj, sc.defaultPruneOption) {
```

agaudreault reviewed (changes requested):
Overall, the feature LGTM. Few refactors required here and there.

agaudreault reviewed (approved):
LGTM. Ready to merge when test are fixed.

MrFreezeex committed:
Merge branch 'master' into sync-option-confirm
